### PR TITLE
Optimizing away the `fill-mask` pipeline.

### LINKED ISSUE #12099: FillMaskPipeline very slow when provided with a large `targets`
## Environment info

- `transformers` version: 4.6.1
- Platform: Linux-5.4.0-67-generic-x86_64-with-glibc2.10
- Python version: 3.8.5
- PyTorch version (GPU?): 1.8.1 (False)
- Tensorflow version (GPU?): N/A
- Using GPU in script?: No
- Using distributed or parallel set-up in script?: No

### Who can help
@LysandreJik @Narsil 

## Information

The model I am using: `ethanyt/guwenbert-base`, with a `RoBERTa` model and a `BertTokenizerFast` tokenizer.

## To reproduce

1. Initialize a `fill-mask` pipeline with the model and the tokenizer mentioned above
2. Call it with any sentence and a large `targets` (with a length of ~10k single words)

## Problem

The call would be much slower than a similar call without a `targets` argument. A call without a `targets` argument costs ~0.1s, while a call with a `targets` argument costs ~0.3s.

The following code is present in `src/transformers/pipelines/fill_mask.py`:

```python
class FillMaskPipeline(Pipeline):
    # ...
    def __call__(self, *args, targets=None, top_k: Optional[int] = None, **kwargs):
        # ...
        if targets is not None:
            # ...
            targets_proc = []
            for target in targets:
                target_enc = self.tokenizer.tokenize(target)
                # ...
                targets_proc.append(target_enc[0])
```

This function iterates through targets, rather than sending it directly to `tokenize`, which does not utilize the batch processing optimization of `TokenizerFast`s, hence the slow speed.

### PR DESCRIPTION
**Pull request by Narsil**
# What does this PR do?

- Don't send anything to the tokenizer unless needed. Vocab check is much
faster
- Keep BC by sending data to the tokenizer when needed. User handling
warning messages will see performance benefits again
- Make `targets` and `top_k` work together better `top_k` cannot be higher
than `len(targets)` but can be smaller still.
- Actually simplify the `target_ids` in case of duplicate (it can happen 
because we're parsing raw strings)
- Removed useless code to fail on empty strings. It works only if empty 
string is in first position, moved to ignoring them instead.
- Changed the related tests as only the tests would fail correctly
(having incorrect value in first position)

Fixes #12099

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@LysandreJik @EtaoinWu

### CODE REVIEW

File: tests/test_pipelines_fill_mask.py [hunk 1]
```diff
@@ -111,7 +136,7 @@ def test_torch_fill_mask_results(self):
             "My name is <mask>",
             "The largest city in France is <mask>",
         ]
-        valid_targets = [" Patrick", " Clara"]
+        valid_targets = ["ĠPatrick", "ĠClara"]
```
Comment (Narsil) [inline tests/test_pipelines_fill_mask.py:139]: Those tests could stay the same, but it's actually missing from the vocabulary, so it's hitting the slow path and triggering warnings.

the 2 small models don't share the same vocabulary, so if we want tests that avoid the slow path all  the time we should have 2 version of the targets IMO.

**Comment by Narsil** [comment]
Ping @LysandreJik Do you mind doing a quick review ?

(Tests do not have to be modified, for this to work, but it will output a lot of warning and be slower than necessary)

**Comment by LysandreJik** [comment]
Thanks for the ping, reviewing now!

File: src/transformers/pipelines/fill_mask.py [hunk 2]
```diff
@@ -115,25 +115,53 @@ def __call__(self, *args, targets=None, top_k: Optional[int] = None, **kwargs):
...
-                targets_proc.append(target_enc[0])
-            target_inds = np.array(self.tokenizer.convert_tokens_to_ids(targets_proc))
+                target_ids.append(id_)
+            target_ids = list(set(target_ids))
+            if len(target_ids) == 0:
+                raise ValueError("At least one target must be provided when passed.")
+            target_ids = np.array(target_ids)
+            # Cap top_k if there are targets
+            if top_k > target_ids.shape[0]:
+                top_k = target_ids.shape[0]
```
Comment (LysandreJik) [inline src/transformers/pipelines/fill_mask.py:164]: Nice

File: src/transformers/pipelines/fill_mask.py [hunk 3]
```diff
@@ -115,25 +115,53 @@ def __call__(self, *args, targets=None, top_k: Optional[int] = None, **kwargs):
...
         batch_size = outputs.shape[0] if self.framework == "tf" else outputs.size(0)
 
         if targets is not None:
-            if len(targets) == 0 or len(targets[0]) == 0:
-                raise ValueError("At least one target must be provided when passed.")
             if isinstance(targets, str):
                 targets = [targets]
 
-            targets_proc = []
+            vocab = self.tokenizer.vocab
```
Comment (LysandreJik) [inline src/transformers/pipelines/fill_mask.py:129]: Not all tokenizers have the `vocab` property, for example the slow `RobertaTokenizer` doesn't have it

File: src/transformers/pipelines/fill_mask.py [hunk 4]
```diff
@@ -115,25 +115,53 @@ def __call__(self, *args, targets=None, top_k: Optional[int] = None, **kwargs):
...
+            target_ids = []
             for target in targets:
-                target_enc = self.tokenizer.tokenize(target)
-                if len(target_enc) > 1 or target_enc[0] == self.tokenizer.unk_token:
+                id_ = vocab.get(target, None)
```
Comment (LysandreJik) [inline src/transformers/pipelines/fill_mask.py:132]: Here instead of using `vocab.get` which isn't available on all tokenizers, you could encode the target without special tokens and ensure it returns a list with a single value that is not an unknown token.

I can't think of a simpler way that is tokenizer agnostic.

Comment (LysandreJik) [reply src/transformers/pipelines/fill_mask.py:132]: If you still want to keep the speed of `vocab.get`, maybe putting it behind a statement such as `if hasattr(self.tokenizer, "vocab")` would help.

File: src/transformers/pipelines/fill_mask.py [hunk 5]
```diff
@@ -115,25 +115,53 @@ def __call__(self, *args, targets=None, top_k: Optional[int] = None, **kwargs):
...
+                        add_special_tokens=False,
+                        return_attention_mask=False,
+                        return_token_type_ids=False,
+                        max_length=1,
+                        truncation=True,
+                    )["input_ids"]
+                    if len(input_ids) == 0:
+                        logger.warning(
+                            f"The specified target token `{target}` does not exist in the model vocabulary. "
+                            f"We cannot replace it with anything meaningful, irngoring it"
```
Comment (LysandreJik) [inline src/transformers/pipelines/fill_mask.py:145]: Wouldn't this return the unknown token if it doesn't exist in the tokenizer vocab?

Comment (LysandreJik) [reply src/transformers/pipelines/fill_mask.py:145]: For example:

```py
>>> from transformers import BertTokenizerFast
>>> tokenizer = BertTokenizerFast.from_pretrained("bert-base-cased")
>>> tokenizer("珞", add_special_tokens=False, max_length=1, truncation=True)["input_ids"]
[100]
```
with `100` being the unknown token

**Review by LysandreJik — COMMENTED** [review]
Thanks for working on this! I've added a comment related to tokenizer agnosticity.

For some pipelines (e.g., token classif) we put a requirement on using a fast tokenizer so we could eventually require that here too - but I think we can do with both.

File: src/transformers/pipelines/fill_mask.py (hunk 3 above)
Comment (Narsil) [reply src/transformers/pipelines/fill_mask.py:129]: Is there any way to get reliably a vocab (or reverse vocab) ? That undermines the purpose of this PR which is to add fast lookup. `get_vocab()` exists and is mandatory for all fast tokenizers for instance.

In this case we don't really need to `tokenize` but really just a way to get corresponding ids to reduce the potential targets.
The rationale here is:
- Fast lookup
- If lookup fail, slow tokenization and take first token. It will enable more "natural" usage but still most likely slightly wrong.

If "something" does not correspond to a single token ("some", + "thing") but is in the target list, then we will filter on "some" and it will be a possible output of the fill-mask which is .. subpar but best effort.

File: src/transformers/pipelines/fill_mask.py (hunk 4 above)
Comment (Narsil) [reply src/transformers/pipelines/fill_mask.py:132]: That's another option, but I don't think we can beat the lookup.

The `tokenizer` part is actually quite slow (compared to a Python dict lookup) here. Especially when called sequentially.
When called batched it was reduced heavily but still non negligible (something like 30ms for 10k targets).
I would have to dive to make sure, but I think it's linked to the strings which have to all be copied and parsed at the Python/Rust interface.

There are other implementations which  I tried that still speedup compared to master, but I had the impression to do anything more than raw lookup for best case scenario was wrong. That's actually what the doc says too, the `targets` should be in the vocab.

File: src/transformers/pipelines/fill_mask.py (hunk 5 above)
Comment (Narsil) [reply src/transformers/pipelines/fill_mask.py]: Not necessarily for the empty string "". (I guess single space could give the same if stripping is part of the tokenization preprocessing step, I am not sure)

File: src/transformers/pipelines/fill_mask.py (hunk 3 above)
Comment (Narsil) [reply src/transformers/pipelines/fill_mask.py:129]: @LysandreJik just checked, `get_vocab` seems like supported everywhere, isn't it ? (It's at least the case for Roberta)

I can still try and fail to err on the side of caution.

Comment (LysandreJik) [reply src/transformers/pipelines/fill_mask.py]: Yes, indeed, the `get_vocab` is tested for all tokenizers! Let's go with that one.

File: src/transformers/pipelines/fill_mask.py [hunk 6]
```diff
@@ -115,25 +115,56 @@ def __call__(self, *args, targets=None, top_k: Optional[int] = None, **kwargs):
...
         if targets is not None:
-            if len(targets) == 0 or len(targets[0]) == 0:
-                raise ValueError("At least one target must be provided when passed.")
             if isinstance(targets, str):
                 targets = [targets]
 
-            targets_proc = []
+            try:
+                vocab = self.tokenizer.get_vocab()
+            except Exception:
```
Comment (LysandreJik) [inline src/transformers/pipelines/fill_mask.py:131]: Do you think we could have a better exception than this?

**Review by LysandreJik — APPROVED** [review]
Ok yes, LGTM!

This makes me think that the pipelines should always be tested with multiple models and tokenizers. It would make the development easier while ensuring we do not miss any odd cases.

I tested it out locally with the following code:

```py
from transformers import MODEL_FOR_MASKED_LM_MAPPING, TOKENIZER_MAPPING, pipeline
from huggingface_hub import HfApi

api = HfApi()

for config_class in MODEL_FOR_MASKED_LM_MAPPING.keys():
    try:
        print(config_class)
        model_class, tokenizer_classes = MODEL_FOR_MASKED_LM_MAPPING[config_class], TOKENIZER_MAPPING[config_class]
        model_type = config_class.model_type

        checkpoints = api.list_models(filter=model_type)
        checkpoint = checkpoints[0]

        model = model_class(config_class.from_pretrained(checkpoint.modelId))

        for tokenizer_class in [t for t in tokenizer_classes if t is not None]:
            tokenizer = tokenizer_class.from_pretrained(checkpoint.modelId)

            masker = pipeline("fill-mask", model=model, tokenizer=tokenizer)
            masker(f"Hey who is {tokenizer.mask_token} person?")
            masker(f"Hey who is {tokenizer.mask_token} person?", targets="hey")
            masker(f"Hey who is {tokenizer.mask_token} person?", targets=["hey"])

    except Exception as e:
        print(f"Couldn't get it to work with for", config_class.__name__, ". Error:", e)
```

I manually verified that all failures were either bogus, or not working before either (which isn't surprising, for models such as Wav2Vec2)

File: src/transformers/pipelines/fill_mask.py (hunk 6 above)
Comment (Narsil) [reply src/transformers/pipelines/fill_mask.py:131]: Depends if we want to be permissive or restrictive in the type of error we catch.
Ideally there would not be a try except. But I was not sure it was mandatory to have such a method.

If we consider this could fail for valid reasons (not implemented), I don't really think we should make assumptions about the type of exceptions. NotImplementedError, AttributeError, but it could very well be a signature issue or something else.

I think we should either keep it like this (very permissive in types of errors) or remove the try altogether and be very restrictive (tokenizers MUST implement a valid `get_vocab`)

**Comment by Narsil** [comment]
That's a very neat idea !

It must be quite slow though, right ?

File: src/transformers/pipelines/fill_mask.py (hunk 6 above)
Comment (LysandreJik) [reply src/transformers/pipelines/fill_mask.py:131]: Alright, works for me!

**Comment by LysandreJik** [comment]
Yes, definitely too slow to actually put in tests and generally a bad idea to rely on model hub checkpoints for this I think, it was just the quickest way to ensure that all tokenizer/model pairs really do continue working

**Comment by Narsil** [comment]
Yes, maybe have a script or something for larger refactors for sure.
